src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,23 +124,102 @@ Score(const size_t queryIndex, TreeType& referenceNode)
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
+
+    // Store not used alpha for Monte Carlo.
+    if (kernelIsGaussian && monteCarlo)
+      accumMCAlpha(queryIndex) += CalculateAlpha(referenceNode.Stat().Depth());
+  }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= mcAccessCoef * initialSampleSize &&
+           kernelIsGaussian)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    // Calculate alpha.
+    const double depthAlpha = CalculateAlpha(referenceNode.Stat().Depth());
+    const double alpha = depthAlpha + accumMCAlpha(queryIndex);
+
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, alpha / 2));
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    size_t m = initialSampleSize;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+    while (m > 0)
+    {
+      const size_t oldSize = sample.size();
+      const size_t newSize = oldSize + m;
+
+      // Don't use probabilistic estimation if this is gonna take a close
+      // amount of computation to the exact calculation.
+      if (newSize >= mcBreakCoef * numDesc)
+      {
+        useMonteCarloPredictions = false;
+        break;
+      }
+
+      // Increase the sample size.
+      sample.resize(newSize);
+      for (size_t i = 0; i < m; ++i)
+      {
+        // Sample and evaluate random points from the reference node.
+        const size_t randomPoint = math::RandInt(0, numDesc);
+        sample(oldSize + i) =
+            EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+      }
+      meanSample = arma::mean(sample);
+      const double stddev = arma::stddev(sample);
+      const double mThreshBase =
+          z * stddev * (1 + relError) / (relError * meanSample);
+      const size_t mThresh = std::ceil(mThreshBase * mThreshBase);
+
+      if (sample.size() < mThresh)
+        m = mThresh - sample.size();
+      else
+        m = 0;
+    }
+
+    if (useMonteCarloPredictions)
+    {
+      // Use Monte Carlo estimation.
+      accumMCAlpha(queryIndex) = 0;
+      score = DBL_MAX;
+      densities(queryIndex) += numDesc * meanSample;
+    }
+    else
+    {
+      if (referenceNode.IsLeaf())
+      {
+        // Reclaim not used alpha since the node will be exactly computed.
+        accumMCAlpha(queryIndex) +=
+            CalculateAlpha(referenceNode.Stat().Depth());
```

Comment:
Hmm, but it's possible that non-Monte Carlo approximation may occur at a non-leaf level, meaning that that node's unused probability isn't actually ever added to `accumMCAlpha(queryIndex)`.

Reply:
I think this is fixed for single-trees in f0aa09d651007b0c12f7b01f724f959f776b3e52.

Reply:
And also for dual-trees in e4d507d4507218d5c2051aa538e9014c587c9dd8 :)

Reply:
Nice.  Actually I thought this code would be really difficult and complex, but now that I read through what you did it turns out to be pretty simple. :+1: